File: balanced_brackets.rb
#!usr/bin/env ruby
# http://rosettacode.org/wiki/Balanced_brackets

def balanced(s,index=0,balance=0)
  if balance < 0
    return false
  elsif index == s.length
    return balance == 0
  elsif s[index] == "["
    balanced(s,index+1,balance+1)
  elsif s[index] == "]"
    balanced(s,index+1,balance-1)
  else
    balances(s,index+1,balance)
  end
end


zero = "[" # false
one = "[]" # true
two = "][" # false
three = "[][]" # true
four = "][][" # false
five = "[[][]]" # true
six = "[]][[]" # false

[zero,one,two,three,four,five,six].each { |i| p balanced(i) }

File: balanced_brackets.py
#!/usr/bin/env python
# http://rosettacode.org/wiki/Balanced_brackets

def balanced(s):
  def _bal_brackets(s,index,balance):
    if balance < 0:
      return False
    elif index == len(s):
      return balance == 0
    elif s[index] == "[":
      return _bal_brackets(s,index+1,balance+1)
    elif s[index] == "]":
      return _bal_brackets(s,index+1,balance-1)
    else:
      return _bal_brackets(s,index+1,balance)
  return _bal_brackets(s,0,0)

zero = "[" # false
one = "[]" # true
two = "][" # false
three = "[][]" # true
four = "][][" # false
five = "[[][]]" # true
six = "[]][[]" # false

for e in [zero,one,two,three,four,five,six]:
  print balanced(e)